File: desktop/ProtonEcuToolkit.Core/Transport/SerialTransport.cs
using System.IO.Ports;
using System.Text;

namespace ProtonEcuToolkit.Core.Transport;

/// <summary>
/// Raw serial framing only. Writes a line, reads bytes until the ELM327's
/// '&gt;' prompt shows up, and hands back whatever text arrived in between.
/// Has no idea what an AT command or a KWP2000 service byte is - that
/// belongs to the layers above.
/// </summary>
public sealed class SerialTransport : IDisposable
{
    private const char Prompt = '>';
    private const int DefaultBaudRate = 38400;

    private readonly string _portName;
    private readonly int _baudRate;
    private readonly StringBuilder _buffer = new();
    private readonly SemaphoreSlim _queue = new(1, 1);
    private SerialPort? _port;
    private PendingRequest? _pending;

    public SerialTransport(string portName, int baudRate = DefaultBaudRate)
    {
        _portName = portName;
        _baudRate = baudRate;
    }

    public static string[] ListPortNames() => SerialPort.GetPortNames();

    public bool IsOpen => _port?.IsOpen ?? false;

    public Task OpenAsync()
    {
        var port = new SerialPort(_portName, _baudRate);
        port.DataReceived += OnDataReceived;
        port.ErrorReceived += (_, _) => FailPending(new IOException("SerialTransport: port error"));
        port.Open();
        _port = port;
        return Task.CompletedTask;
    }

    public Task CloseAsync()
    {
        if (_port is { IsOpen: true } port)
        {
            port.DataReceived -= OnDataReceived;
            port.Close();
        }
        _port = null;
        return Task.CompletedTask;
    }

    /// <summary>
    /// Writes one line (CR-terminated) and resolves with everything received
    /// before the next '&gt;' prompt. Calls are serialized - only one in flight
    /// at a time, since the ELM327 is strictly half-duplex.
    /// </summary>
    public async Task<string> SendAsync(string line, int timeoutMs = 2000)
    {
        await _queue.WaitAsync().ConfigureAwait(false);
        try
        {
            return await SendNowAsync(line, timeoutMs).ConfigureAwait(false);
        }
        finally
        {
            _queue.Release();
        }
    }

    private Task<string> SendNowAsync(string line, int timeoutMs)
    {
        var port = _port;
        if (port is not { IsOpen: true })
        {
            throw new InvalidOperationException("SerialTransport: port is not open");
        }

        lock (_buffer)
        {
            _buffer.Clear();
        }

        var tcs = new TaskCompletionSource<string>(TaskCreationOptions.RunContinuationsAsynchronously);
        var cts = new CancellationTokenSource(timeoutMs);
        var pending = new PendingRequest(tcs, cts);
        cts.Token.Register(() =>
        {
            if (Interlocked.CompareExchange(ref _pending, null, pending) == pending)
            {
                tcs.TrySetException(
                    new TimeoutException($"SerialTransport: timed out waiting for response to \"{line}\""));
            }
        });

        _pending = pending;
        port.Write(line + "\r");
        return tcs.Task;
    }

    private void OnDataReceived(object sender, SerialDataReceivedEventArgs e)
    {
        var port = (SerialPort)sender;
        string chunk;
        try
        {
            chunk = port.ReadExisting();
        }
        catch (Exception)
        {
            return;
        }

        string? response = null;
        lock (_buffer)
        {
            _buffer.Append(chunk);
            if (_pending is null) return;

            var text = _buffer.ToString();
            var promptIndex = text.IndexOf(Prompt);
            if (promptIndex == -1) return;

            response = text[..promptIndex];
            _buffer.Clear();
            _buffer.Append(text[(promptIndex + 1)..]);
        }

        var pending = Interlocked.Exchange(ref _pending, null);
        if (pending is null) return;
        pending.Cts.Dispose();
        pending.Tcs.TrySetResult(response);
    }

    private void FailPending(Exception ex)
    {
        var pending = Interlocked.Exchange(ref _pending, null);
        if (pending is null) return;
        pending.Cts.Dispose();
        pending.Tcs.TrySetException(ex);
    }

    public void Dispose()
    {
        if (_port is { IsOpen: true })
        {
            _port.Close();
        }
        _port?.Dispose();
        _queue.Dispose();
    }

    private sealed record PendingRequest(TaskCompletionSource<string> Tcs, CancellationTokenSource Cts);
}

File: src/server/transport/SerialTransport.ts
import { SerialPort } from "serialport";

export interface SerialTransportOptions {
  path: string;
  baudRate?: number;
}

const PROMPT = ">";
const DEFAULT_BAUD_RATE = 38400;

/**
 * Raw serial framing only. Writes a line, reads bytes until the ELM327's
 * '>' prompt shows up, and hands back whatever text arrived in between.
 * Has no idea what an AT command or a KWP2000 service byte is - that
 * belongs to the layers above.
 */
export class SerialTransport {
  private port: SerialPort | null = null;
  private buffer = "";
  private pending: {
    resolve: (value: string) => void;
    reject: (reason: Error) => void;
    timer: NodeJS.Timeout;
  } | null = null;
  private queue: Promise<void> = Promise.resolve();

  constructor(private readonly options: SerialTransportOptions) {}

  static listPorts() {
    return SerialPort.list();
  }

  get isOpen(): boolean {
    return this.port?.isOpen ?? false;
  }

  open(): Promise<void> {
    return new Promise((resolve, reject) => {
      const port = new SerialPort(
        {
          path: this.options.path,
          baudRate: this.options.baudRate ?? DEFAULT_BAUD_RATE,
          autoOpen: false,
        },
        (err) => {
          if (err) reject(err);
        },
      );
      port.on("data", (chunk: Buffer) => this.onData(chunk));
      port.on("error", (err) => this.failPending(err));
      port.on("close", () => this.failPending(new Error("SerialTransport: port closed")));
      port.open((err) => {
        if (err) {
          reject(err);
          return;
        }
        this.port = port;
        resolve();
      });
    });
  }

  close(): Promise<void> {
    return new Promise((resolve, reject) => {
      if (!this.port?.isOpen) {
        resolve();
        return;
      }
      this.port.close((err) => {
        if (err) reject(err);
        else resolve();
      });
    });
  }

  /**
   * Writes one line (CR-terminated) and resolves with everything received
   * before the next '>' prompt. Calls are serialized - only one in flight
   * at a time, since the ELM327 is strictly half-duplex.
   */
  send(line: string, timeoutMs = 2000): Promise<string> {
    const previous = this.queue;
    let release!: () => void;
    this.queue = new Promise<void>((res) => {
      release = res;
    });
    return previous.then(() => this.sendNow(line, timeoutMs)).finally(() => release());
  }

  private sendNow(line: string, timeoutMs: number): Promise<string> {
    const port = this.port;
    if (!port?.isOpen) {
      return Promise.reject(new Error("SerialTransport: port is not open"));
    }
    return new Promise((resolve, reject) => {
      this.buffer = "";
      const timer = setTimeout(() => {
        this.pending = null;
        reject(new Error(`SerialTransport: timed out waiting for response to "${line}"`));
      }, timeoutMs);
      this.pending = { resolve, reject, timer };
      port.write(`${line}\r`, (err) => {
        if (err) this.failPending(err);
      });
    });
  }

  private onData(chunk: Buffer): void {
    this.buffer += chunk.toString("utf8");
    if (!this.pending) return;
    const promptIndex = this.buffer.indexOf(PROMPT);
    if (promptIndex === -1) return;
    const response = this.buffer.slice(0, promptIndex);
    this.buffer = this.buffer.slice(promptIndex + 1);
    const { resolve, timer } = this.pending;
    clearTimeout(timer);
    this.pending = null;
    resolve(response);
  }

  private failPending(err: Error): void {
    if (!this.pending) return;
    clearTimeout(this.pending.timer);
    const { reject } = this.pending;
    this.pending = null;
    reject(err);
  }
}
